### ISSUE TITLE
Nomad-exporter transiently fails due to allocation metric collission

### ISSUE THREAD

**Issue by dgonzalezruiz** [pcarranza/nomad-exporter#18]
We realised that nomad-exporter failed with some level of frequency, specially after starting to use nodes that allowed higher levels of "container density"; it started returning 500 upon scrape. The response contained several lines, flooded with the same thing:

```
* collected metric nomad_allocation_cpu_system_mode label:<name:"alloc" value:"oosomjob.oosomejob[0]" > label:<name:"datacenter" value:"oosomdc" > label:<name:"group" value:"oosomjob" > label:<name:"job" value:"oosomjob" > label:<name:"node" value:"oosomnode" > label:<name:"region" value:"global" > gauge:<value:380.915828904299 >  was collected before with the same name and label values
```

This happened because a canary allocation, labeled as above, was scheduled in the same node as the canary of the previous deployment of the job.
Note that canaries will always be labelled to have `alloc` = `XXXX.YYYY[0]`, so a collision happens sometimes.
The same thing will happen if the 2nd,3rd,4th.. allocations of the job are scheduled in the same node as the previous allocation of the deployment.

After removing the allocation(s) introducing this collision on `nomad-exporter`, the exporter stops responding `500`. 

I believe we should add an extra label with `alloc_id` in order to introduce a unique identifier to the metrics. Given that we are already filtering out metrics of stale allocations, this shouldn't introduce a cardinality bomb. 

WDYT @pcarranza ?

**Comment by pcarranza** [comment]
Nah, we need a way of identifying the canary instead. That way we will not be introducing an ID which can derive into a bomb.

**Comment by dgonzalezruiz** [comment]
Note that this not only affects canaries; the `alloc` label has `XXXX.YYYY[N]` possible values, where `N`:`[0, $job_redundancy)`. If any of these ordered elements happens to be scheduled in the same  node as in the case of the previous deployment (i.e. not a canary, but another alloc in the middle of an ongoing deployment), this will still cause a collision and bring back error responses.

**Comment by pcarranza** [comment]
Gotcha, then it's not about canaries, but about the allocation being in the same place with the same name and index in the list.

Then the difference will be the version, or the deployment id. Would that make sense?

**Comment by pcarranza** [comment]
@dgonzalezruiz I've something for you to test.

I moved the development to gitlab to have some actual CI, you can now find it here: https://gitlab.com/yakshaving.art/nomad-exporter

This adds a _job_version_ to allocations and deployments, removes the _deployment_id_ to reduce cardinality (not a lot, but meh, it does make sense) and _failed_ from the task count because it was basically a useless metric.

Though, if you were running this as a binary I will recommend you to run it as a container within the nomad cluster, this way you can control memory usage (to prevent it from burning all the available memory) - you can find a nomad-exporter.nomad hcl file which you can use as a baseline.

I've been testing it by installing multiple versions of the same thing (the nomad exporter itself) in a single node, and I could not see any collision.

Please, try it, if it works fine I'll tag it to lock the version.
